### PR TITLE
Fix bug when running presence off master

### LINKED ISSUE #10027: Own presence shown as offline after a while with presence writer worker enabled
### Description

We're trialing presence workers on a few EMS hosts and for a while the new presence writer worker worked fine. After a few days, both of the hosts stopped reporting presence online for the user themselves. A restart of Synapse made presence show correctly for self again.

### Version information

- **Version**: v1.34.0
- **Platform**: EMS/Docker

Please ping for host names for logs access.

### PR DESCRIPTION
**Pull request by erikjohnston**
Hopefully fixes #10027.

### CODE REVIEW

**Review by callahad — COMMENTED** [review]
Seems reasonable. Two completely subjective questions:

- How do we feel about `assert`, versus code which is guaranteed to execute at runtime?

- Given the duplication, would it make sense to throw the assertion in a decorator and slap that on the `get_next` functions?

**Comment by anoadragon453** [comment]
> How do we feel about assert, versus code which is guaranteed to execute at runtime?

I've always been a bit wary about it, as its possible to strip `assert`s out when running Python with `-O` and `-OO`, which some people may be doing. Alternatively, I think using them only in the context of when a programming error was made (assert that a str was passed to this function) is fine.

I'm not sure if this change falls into the latter bracket though, so perhaps raising an Exception would be better?

File: synapse/storage/util/id_generators.py [hunk 1]
```diff
@@ -397,6 +397,11 @@ def get_next(self):
                 # ... persist event ...
         """
 
+        # If we have a list of instance that are allowed to write to this
```
Comment (richvdh) [inline synapse/storage/util/id_generators.py:400]:
```suggestion
        # If we have a list of instances that are allowed to write to this
```

File: synapse/storage/util/id_generators.py [hunk 2]
```diff
@@ -406,6 +411,11 @@ def get_next_mult(self, n: int):
                 # ... persist events ...
         """
 
+        # If we have a list of instance that are allowed to write to this
```
Comment (richvdh) [inline synapse/storage/util/id_generators.py:414]:
```suggestion
        # If we have a list of instances that are allowed to write to this
```

File: synapse/storage/util/id_generators.py [hunk 3]
```diff
@@ -416,6 +426,11 @@ def get_next_txn(self, txn: LoggingTransaction):
             # ... persist event ...
         """
 
+        # If we have a list of instance that are allowed to write to this
```
Comment (richvdh) [inline synapse/storage/util/id_generators.py:429]:
```suggestion
        # If we have a list of instances that are allowed to write to this
```

File: synapse/storage/databases/main/presence.py [hunk 4]
```diff
@@ -50,7 +50,7 @@ def __init__(
                 instance_name=self._instance_name,
                 tables=[("presence_stream", "instance_name", "stream_id")],
                 sequence_name="presence_stream_sequence",
-                writers=hs.config.worker.writers.to_device,
+                writers=hs.config.worker.writers.presence,
```
Comment (richvdh) [inline synapse/storage/databases/main/presence.py:53]: hawkward!

File: changelog.d/10149.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Fix bug where presence updates would stop working when using a presence writer worker.
```
Comment (richvdh) [inline changelog.d/10149.bugfix:1]:
```suggestion
Fix a bug which caused  presence updates to stop working some time after restart, when using a presence writer worker.
```

**Review by richvdh — APPROVED** [review]
lgtm

File: synapse/storage/databases/main/presence.py (hunk 4 above)
Comment (erikjohnston) [reply synapse/storage/databases/main/presence.py:53]: inorite

**Pull request merged**